java-checks/src/main/java/org/sonar/java/checks/synchronization/WriteObjectTheOnlySynchronizedMethodCheck.java

```diff
@@ -0,0 +1,91 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.synchronization;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.serialization.SerializableContract;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.List;
+
+import static org.sonar.java.model.ModifiersUtils.hasModifier;
+import static org.sonar.plugins.java.api.tree.Modifier.SYNCHRONIZED;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.METHOD;
+
+@Rule(key = "S3042")
+public class WriteObjectTheOnlySynchronizedMethodCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodTree methodTree = (MethodTree) tree;
+    Symbol.TypeSymbol enclosingClass = methodTree.symbol().enclosingClass();
+    String className = enclosingClass.type().fullyQualifiedName();
+    MethodMatcher writeObjectMatcher = SerializableContract.writeObjectMatcher(className);
+    if (writeObjectMatcher.matches(methodTree) && hasModifier(methodTree.modifiers(), SYNCHRONIZED)) {
+      SynchronizationVisitor visitor = new SynchronizationVisitor(methodTree);
+      enclosingClass.declaration().accept(visitor);
+      if (!visitor.moreThanSingleLock) {
+        reportIssue(ModifiersUtils.getModifier(methodTree.modifiers(), SYNCHRONIZED), "Remove this \"synchronized\" keyword.");
+      }
+    }
+  }
+
+  private static class SynchronizationVisitor extends BaseTreeVisitor {
+
+    private final MethodTree writeObjectMethodTree;
+    private boolean moreThanSingleLock;
+
+    public SynchronizationVisitor(MethodTree writeObjectMethodTree) {
+      this.writeObjectMethodTree = writeObjectMethodTree;
+    }
+
+    @Override
+    public void visitSynchronizedStatement(SynchronizedStatementTree tree) {
+      moreThanSingleLock = true;
+    }
+
+    @Override
+    public void visitMethod(MethodTree tree) {
+      if (tree == writeObjectMethodTree || moreThanSingleLock) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)